skstore/src/Extern.sk

```diff
@@ -0,0 +1,668 @@
+  )
+    .getDirName()
+    .toString()
+}
+
+@export("SKIP_SKStore_multimap")
+fun multimap(
+  context: mutable Context,
+  name: String,
+  mappings: SKJSON.CJArray,
+): String {
+  skmappings = mappings match {
+  | SKJSON.CJArray(v) ->
+    v.map(e -> {
+      mapping = SKJSON.expectArray(e);
+      handleId = SKJSON.asString(mapping[0]);
+      eager = EHandle(
+        JSONID::keyType,
+        JSONFile::type,
+        DirName::create(handleId),
+      );
+      mapHandle = FnHandle::make(
+        UInt32::truncate(SKJSON.asFloat(mapping[1]).toInt()),
+      );
+      return (eager, mapHandle.map)
+    })
+  };
+  eagerMultiMap(JSONID::keyType, JSONFile::type, name, context, skmappings)
+    .getDirName()
+    .toString()
+}
+
+@export("SKIP_SKStore_multimapReduce")
+fun multimapReduce(
+  context: mutable Context,
+  name: String,
+  mappings: SKJSON.CJArray,
+  accumulator: UInt32,
+  default: SKJSON.CJSON,
+): String {
+  accumulatorHandle = FnHandle::make(accumulator);
+  skmappings = mappings match {
+  | SKJSON.CJArray(v) ->
+    v.map(e -> {
+      mapping = SKJSON.expectArray(e);
+      handleId = SKJSON.asString(mapping[0]);
+      eager = EHandle(
+        JSONID::keyType,
+        JSONFile::type,
+        DirName::create(handleId),
+      );
+      mapHandle = FnHandle::make(
+        UInt32::truncate(SKJSON.asFloat(mapping[1]).toInt()),
+      );
+      return (eager, mapHandle.map)
+    })
+  };
+  eagerMultiMapReduce(
+    JSONID::keyType,
+    JSONFile::type,
+    name,
+    context,
+    skmappings,
+    ExternAccumulator(accumulatorHandle, JSONFile(default)),
+  )
+    .getDirName()
+    .toString()
+}
+
+class AsyncValueSaveError(
+  key: SKJSON.CJSON,
+  params: SKJSON.CJSON,
+  reason: String,
+) extends .Exception {
+  fun getMessage(): String {
+    `Unable to save async result for key '${this.key.toJSON()}' and params '${this.params.toJSON()}: ${
+      this.reason
+    }'.`
+  }
+}
+
+@export("SKIP_SKStore_asyncResult")
+fun asyncResult(
+  callId: String,
+  lazyName: String,
+  key: SKJSON.CJSON,
+  params: SKJSON.CJSON,
+  value: SKJSON.CJObject,
+): Float {
+  SKStore.runLockedWithResult(context ~> {
+    resultsDirs = DirName::create(lazyName + "sk_prv/results/");
+    context.unsafeMaybeGetEagerDir(resultsDirs) match {
+    | Some(dir) ->
+      if (!dir.isInput()) {
+        throw AsyncValueSaveError(key, params, "Invalid async handle")
+      };
+      keyId = JSONID(key);
+      current = dir.getArray(context, keyId).map(JSONFile::type);
+      if (current.isEmpty()) {
+        throw AsyncValueSaveError(key, params, "Query info not found")
+      };
+      cvalue = current[0].value match {
+      | x @ SKJSON.CJObject _ -> x
+      | _ -> throw AsyncValueSaveError(key, params, "Invalid query info")
+      };
+      ksuid = SKJSON.getString(cvalue, "ksuid");
+      if (!ksuid.map(v -> callId == v).default(false)) {
+        // The received response is deprecated
```

Comment:
The response is deprecated because new lazy computation was fired  for the concerned key so the current response is useless because it correspond to an old state.